cdap-app-fabric/src/main/java/co/cask/cdap/internal/app/namespace/NamespaceService.java

```diff
@@ -0,0 +1,215 @@
+/*
+ * Copyright © 2015 Cask Data, Inc.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License"); you may not
+ * use this file except in compliance with the License. You may obtain a copy of
+ * the License at
+ *
+ * http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS, WITHOUT
+ * WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied. See the
+ * License for the specific language governing permissions and limitations under
+ * the License.
+ */
+
+package co.cask.cdap.internal.app.namespace;
+
+import co.cask.cdap.app.store.Store;
+import co.cask.cdap.app.store.StoreFactory;
+import co.cask.cdap.common.conf.Constants;
+import co.cask.cdap.common.exception.AlreadyExistsException;
+import co.cask.cdap.common.exception.NotFoundException;
+import co.cask.cdap.config.DashboardStore;
+import co.cask.cdap.config.PreferencesStore;
+import co.cask.cdap.proto.Id;
+import co.cask.cdap.proto.NamespaceMeta;
+import com.google.common.util.concurrent.AbstractIdleService;
+import com.google.common.util.concurrent.FutureCallback;
+import com.google.common.util.concurrent.Futures;
+import com.google.common.util.concurrent.ListenableFuture;
+import com.google.common.util.concurrent.ListeningExecutorService;
+import com.google.common.util.concurrent.MoreExecutors;
+import com.google.inject.Inject;
+import com.google.inject.Singleton;
+import com.sun.istack.Nullable;
+import org.apache.twill.common.Threads;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+import java.util.List;
+import java.util.concurrent.Callable;
+import java.util.concurrent.Executors;
+import java.util.concurrent.TimeUnit;
+
+/**
+ * {@link AbstractIdleService} for managing namespaces
+ */
+@Singleton
+public final class NamespaceService extends AbstractIdleService {
+  private static final Logger LOG = LoggerFactory.getLogger(NamespaceService.class);
+  private static final String NAMESPACE_ELEMENT_TYPE = "Namespace";
+
+  private final Store store;
+  private final PreferencesStore preferencesStore;
+  private final DashboardStore dashboardStore;
+
+  @Inject
+  public NamespaceService(StoreFactory storeFactory, PreferencesStore preferencesStore, DashboardStore dashboardStore) {
+    this.store = storeFactory.create();
```

Comment:
This should be done in the startup, since `storeFactory.create()` will create the underlying table (which might fail due to various reasons that need retry, and retry shouldn't be done in constructor).